File: alter-ego-mobile/src/services/focus.ts
/**
 * Focus tab API service.
 */

import { apiClient } from "@/services/api";

// ── Types ──────────────────────────────────────────────────────────────────

export type FocusMode = "pomodoro" | "deep_work" | "stopwatch";

export interface FocusTag {
  id: string;
  name: string;
  color: string;
  session_count: number;
  created_at: string;
}

export interface FocusSession {
  id: string;
  mode: FocusMode;
  tag_name?: string | null;
  tag_color?: string | null;
  focus_seconds: number;
  break_seconds: number;
  rounds_completed: number;
  was_abandoned: boolean;
  started_at: string;
  ended_at: string;
}

export interface FocusTagStat {
  tag_id: string | null;
  name: string;
  color: string;
  focus_seconds: number;
  pct: number;
}

export interface FocusStats {
  today: {
    focus_seconds: number;
    break_seconds: number;
    sessions: number;
  };
  weekly_chart: Array<{ date: string; focus_seconds: number }>;
  week_total_focus_seconds: number;
  all_time: {
    total_focus_seconds: number;
    total_sessions: number;
    longest_session_seconds: number;
    avg_session_seconds: number;
    focus_streak_days: number;
    completion_pct: number;
  };
  by_tag: FocusTagStat[];
  recent_sessions: FocusSession[];
}

export interface FocusSettings {
  pomodoro_work_minutes: number;
  pomodoro_short_break_minutes: number;
  pomodoro_long_break_minutes: number;
  pomodoro_rounds: number;
  auto_start_breaks: boolean;
  auto_start_work: boolean;
  sound_enabled: boolean;
}

// ── Helpers ────────────────────────────────────────────────────────────────

export function fmtDuration(seconds: number): string {
  if (seconds < 60) return `${seconds}s`;
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  if (h === 0) return `${m}m`;
  if (m === 0) return `${h}h`;
  return `${h}h ${m}m`;
}

// ── API calls ──────────────────────────────────────────────────────────────

export const fetchFocusTags = (): Promise<FocusTag[]> =>
  apiClient.get<FocusTag[]>("/api/v1/focus/tags").then((r) => (Array.isArray(r) ? r : []));

export const createFocusTag = (name: string, color: string) =>
  apiClient.post<FocusTag>("/api/v1/focus/tags", { name, color });

export const deleteFocusTag = (tagId: string) =>
  apiClient.delete(`/api/v1/focus/tags/${tagId}`);

export interface LogSessionBody {
  mode: FocusMode;
  tag_id?: string | null;
  started_at: string;
  ended_at: string;
  focus_seconds: number;
  break_seconds: number;
  rounds_completed: number;
  was_abandoned: boolean;
}

export const logFocusSession = (body: LogSessionBody) =>
  apiClient.post<{ id: string }>("/api/v1/focus/sessions", body);

export const fetchFocusStats = (): Promise<FocusStats> =>
  apiClient.get<FocusStats>("/api/v1/focus/stats");

export const fetchFocusSettings = (): Promise<FocusSettings> =>
  apiClient.get<FocusSettings>("/api/v1/focus/settings");

export const patchFocusSettings = (updates: Partial<FocusSettings>) =>
  apiClient.patch<FocusSettings>("/api/v1/focus/settings", updates);

File: alter-ego-backend/app/api/focus.py
"""Focus tab API — tags, sessions, stats, settings."""

from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel, Field

from app.api.auth import get_user_id_from_token
from app.services.focus_service import (
    create_tag,
    delete_tag,
    get_focus_settings,
    get_focus_stats,
    get_tags,
    log_session,
    update_focus_settings,
)

router = APIRouter(prefix="/api/v1/focus", tags=["focus"])


# ── Tags ─────────────────────────────────────────────────────────────────────


@router.get("/tags")
async def list_tags(authorization: str = Header(None)):
    user_id = get_user_id_from_token(authorization)
    return await get_tags(user_id)


class CreateTagRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=32)
    color: str = Field(default="#8B5CF6")


@router.post("/tags")
async def create_tag_route(body: CreateTagRequest, authorization: str = Header(None)):
    user_id = get_user_id_from_token(authorization)
    try:
        return await create_tag(user_id, body.name, body.color)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.delete("/tags/{tag_id}")
async def delete_tag_route(tag_id: str, authorization: str = Header(None)):
    user_id = get_user_id_from_token(authorization)
    return await delete_tag(user_id, tag_id)


# ── Sessions ─────────────────────────────────────────────────────────────────


class LogSessionRequest(BaseModel):
    mode: str  # 'pomodoro' | 'deep_work' | 'stopwatch'
    tag_id: str | None = None
    started_at: str  # ISO 8601
    ended_at: str  # ISO 8601
    focus_seconds: int = Field(default=0, ge=0)
    break_seconds: int = Field(default=0, ge=0)
    rounds_completed: int = Field(default=0, ge=0)
    was_abandoned: bool = False


@router.post("/sessions")
async def log_session_route(body: LogSessionRequest, authorization: str = Header(None)):
    user_id = get_user_id_from_token(authorization)
    try:
        return await log_session(
            user_id=user_id,
            mode=body.mode,
            tag_id=body.tag_id,
            started_at=body.started_at,
            ended_at=body.ended_at,
            focus_seconds=body.focus_seconds,
            break_seconds=body.break_seconds,
            rounds_completed=body.rounds_completed,
            was_abandoned=body.was_abandoned,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e


# ── Stats ─────────────────────────────────────────────────────────────────────


@router.get("/stats")
async def get_stats(authorization: str = Header(None)):
    user_id = get_user_id_from_token(authorization)
    return await get_focus_stats(user_id)


# ── Settings ──────────────────────────────────────────────────────────────────


@router.get("/settings")
async def get_settings(authorization: str = Header(None)):
    user_id = get_user_id_from_token(authorization)
    return await get_focus_settings(user_id)


class UpdateSettingsRequest(BaseModel):
    pomodoro_work_minutes: int | None = None
    pomodoro_short_break_minutes: int | None = None
    pomodoro_long_break_minutes: int | None = None
    pomodoro_rounds: int | None = None
    auto_start_breaks: bool | None = None
    auto_start_work: bool | None = None
    sound_enabled: bool | None = None
    vibration_enabled: bool | None = None


@router.patch("/settings")
async def update_settings(body: UpdateSettingsRequest, authorization: str = Header(None)):
    user_id = get_user_id_from_token(authorization)
    updates = {k: v for k, v in body.model_dump().items() if v is not None}
    return await update_focus_settings(user_id, updates)
